Add the AMP guide for Next.js (#1705)
GoogleChrome/web.dev: src/site/content/en/react/how-to-use-amp-in-nextjs-and-why/index.md

---
layout: post

> title: How to use AMP in Next.js (and why)

Reviewer: Love the title 👍

authors:
  - houssein
subhead: |
  Add AMP to any of your Next.js pages to guarantee fast page loads
date: 2019-10-17
draft: true

> ---

Author: The page is now ~1250 words. We usually try to cap pages at ~1000 words. I'll do a pass later to see if anything can be removed, but otherwise I'm fine with it being a little long.

[AMP](https://amp.dev) (Accelerated Mobile Pages) is a web component framework that guarantees

> that your pages will load fast. [Next.js][intro] has built-in support for AMP.

Reviewer: s/"that guarantees that your pages will load fast"/"that guarantees fast page loads"?

> that your pages will load fast. [Next.js][intro] has built-in support for AMP.

Author: ```suggestion
that guarantees fast page loads. [Next.js][intro] has built-in support for AMP.
```

## What will you learn?

This guide first briefly describes [how AMP guarantees fast page loads](#overview), then lets you
try out [the two ways that you can use AMP in a Next.js app](#strategies), then helps you [decide
which approach is best for you](#guidance).

The intended audience for this guide is a web developer who has decided to use Next.js but is

> unsure of whether to support AMP. 

Author: Aside from helping readers evaluate whether this page is right for them, being explicit about the audience also helps us (the web.dev maintainers) remember the original purpose of this document.

We might want to go a step further and define our audience as a web developer who has decided to use Next.js and **considers load performance a high priority**. Because load performance optimization is the unifying theme behind all of our Next.js content.

Reviewer: Love it, definitely helps to be more clear 👍 

I'm okay with both!

{% Aside %}
  This guide was not written for web developers who have decided to use AMP but are unsure of what
  framework to use. We will note briefly however that Next.js could be a good choice because it
  supports [AMP server-side rendering](https://amp.dev/documentation/guides-and-tutorials/optimize-and-measure/server-side-rendering/)
  and makes it easy to serve AMP content without introducing a lot of complexity into your codebase.
{% endAside %}

## How AMP guarantees fast page loads {: #overview }

AMP has two main strategies for guaranteeing fastness:

* **AMP HTML**: A restricted form of HTML that makes certain optimizations mandatory and prohibits
  architectural patterns that lead to slowness. See [How AMP works][how] for a high-level
  overview of the optimizations and restrictions.
* **AMP Cache**: A content cache used by some search engines, such as Google and Bing, that uses

>   [prerendering] to speed up page loads. See [Why AMP Caches exist][why] to learn more about the

Reviewer: Link for prerendering missing?

Author: This is Markdown shorthand. When the text matches the link ID, you can just do `[prerendering]` instead of `[prerendering][prerendering]`. Similar to ES6 object shorthand syntax.

  benefits and tradeoffs of the caches and [How does my AMP page get cached?][how] to understand
  how your AMP pages get into the caches.

## How you can use AMP in Next.js {: #strategies }

There are two ways to use AMP in Next.js:

* The [**Hybrid AMP** approach](#hybrid) lets you create an accompanying AMP version of any
  Next.js page.
* The [**AMP-only** approach](#amponly) lets you make AMP the only option for a page.

{% Aside 'caution' %}
  Whenever you serve AMP pages, you won't be able to run React components client-side because
  React components are not valid AMP components.

> {% endAside %}

Author: @alankent in the email thread you said something along these lines. Does this phrasing capture your point?

Reviewer: Might help to make things more explicit about what Next.js actually does? Something like:

`Whenever you serve AMP pages with Next.js, React components do not get rendered on the client. React components are not valid AMP components and Next.js converts everything to valid AMP."

Wording could probably be better, and I'm sure @alankent will have a better suggestion.

Reviewer 2: "converts everything to valid AMP" is not worded correctly - Next.js does not convert anything - the developer has to do the work of generating AMP output. With AMP you are using Next.js as a templating engine on the server to generate AMP pages (and API hosting etc). You are creating a React client app at all. I worry slightly that people might get the wrong idea, but worth broader discussion with others before committing to a path. Need to think about this a little more - is there a use case of tacking AMP on the side of a React application? (My normal mental model is "I want to create an AMP page, Next.js is a useful templating engine to do it.) I think worth discussing / thinking a bit more first to make sure the message is appropriate/useful.

   specified by width and height. Check out [Layout & media queries][layout] to learn more about
   the supported layouts of AMP elements, and [amp-img] to learn more about that element's
   optimizations.
    
1. View the main version of the page again.

   <figure>
     <img src="mainimg.png"
          alt="The main version of the image.">
     <figcaption>
       The main version of the image.
     </figcaption>
   </figure>

1. View the AMP version of the page again.

   <figure>
     <img src="ampimg.png"
          alt="The AMP version of the image.">
     <figcaption>
       The AMP version of the image.
     </figcaption>
   </figure>

### How to create AMP-only pages {: #amponly }

Next.js also supports AMP-only pages. With this approach, a single AMP page is served and rendered
to users and search engines at all times.

1. To render an AMP-only page, change the value of the `amp` property in the config object to `true`.

   ```jsx
   import React from 'react'

   export const config = { amp: true };

   const Home = () => (
     <p>This is an AMP-only page</p>
   );
     
   export default Home;
   ```

> ## How to decide whether to use the Hybrid AMP or AMP-only approach {: #guidance }

Author: Helping developers decide which framework/toolchain/etc. to use will be a big theme in web.dev content over 2020. We might want to move this to a separate page one day. But for now I think it works well as a section in this guide.

If you're serious about load performance, an AMP-only page could be a good way to make sure
that your page gets fast and stays fast. But here's the catch: in order to guarantee fastness,
AMP must prohibit certain architectural patterns and HTML elements that often lead to slow pages.
For example, AMP doesn't allow custom synchronous JavaScript because 

> [render-blocking resources][blockers] are a common cause of slow page loads.

Reviewer: Wow this reads so much better 🚀

In order to understand whether an AMP-only approach is right for you, you need to figure out
whether all of your frontend code can be represented in AMP HTML:

* Read [How AMP works](https://amp.dev/about/how-amp-works/) to understand AMP's high-level
  architectural restrictions and mandatory optimizations.
* Read [HTML Tags][tags] to see what HTML tags AMP allows and prohibits, and
  browse the [AMP component catalogue](https://amp.dev/documentation/components/) to see the

>   custom components that the AMP community has built to solve common use cases.

Reviewer 3: Add: Take a look at [amp-script](https://amp.dev/documentation/components/amp-script/) to learn how to use custom JavaScript to implement features currently not supported by AMP.

> Even if an AMP-only approach won't work for your page1000, it might still be a good idea to

Reviewer: Most likely a deliberate marker that you placed, but just in case you didn't:

```suggestion
Even if an AMP-only approach won't work for your page, it might still be a good idea to
```

use AMP whenever possible, because of its guaranteed fastness. The Hybrid AMP approach

> provides a way to conditionally serve AMP without introducing a lot of complexity

Reviewer 3: I stumbled over:

> The Hybrid AMP approach
provides a way to conditionally serve AMP without introducing a lot of complexity
into your codebase.

It's correct, but I'd make it more explicit by: `s/The Hybrid AMP approach/The Hybrid AMP approach in Next.js`. In general, hybrid AMP means higher maintenance costs.

into your codebase.

## Next steps

Check out the rest of the guides in [web.dev's Next.js collection][collection] to discover
other ways that you can optimize your Next.js app.

[intro]: /performance-as-a-default-with-nextjs
[layout]: https://amp.dev/documentation/guides-and-tutorials/develop/style_and_layout/control_layout/
[collection]: /react#nextjs
[hook]: https://reactjs.org/docs/hooks-overview.html
[prerendering]: https://developers.googleblog.com/2019/08/the-speed-benefit-of-amp-prerendering.html
[tags]: https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml/#html-tags
[blockers]: /render-blocking-resources
[why]: https://blog.amp.dev/2017/01/13/why-amp-caches-exist/
[how]: https://amp.dev/documentation/guides-and-tutorials/learn/amp-caches-and-cors/how_amp_pages_are_cached/#how-does-my-amp-page-get-cached?
[layout]: https://amp.dev/documentation/guides-and-tutorials/develop/style_and_layout/control_layout/

> [amp-img]: https://amp.dev/documentation/examples/components/amp-img/

Reviewer: LOVE the entire outro
